rslib/src/storage/card/mod.rs

```diff
@@ -249,7 +249,7 @@ impl super::SqliteStorage {
         F: FnMut(NewCard) -> bool,
     {
         let mut stmt = self.db.prepare_cached(&format!(
-            "{} ORDER BY due ASC",
+            "{} ORDER BY due, ord ASC",
```

Comment:
I don't understand why this is required for the new test to pass. Shouldn't the cards (siblings) of the grandchild deck be in the right order anyway?

Either way, this also eliminates the nasty special case handled in `add_new_card()`. Is there any reason why the index isn't on (did, queue, due, ord)? Then this new query should be possible without any slowdown.

Reply:
> I don't understand why this is required for the new test to pass. Shouldn't the cards (siblings) of the grandchild deck be in the right order anyway?

`select * from cards` should (but as far as I'm aware, is not guaranteed to) deliver cards in card id order, as it's walking the primary key id btree. In the common case, that will also deliver cards in template order, but that is not always the case (such as when some siblings are generated later). 

In this case though we're relying on a separate index, so presumably the cards are coming back in order that index happens to be stored in. In the case of identical due values, sqlite would be free to store them in either order, so that's probably where the unpredictability is coming from.
 
> Either way, this also eliminates the nasty special case handled in add_new_card(). Is there any reason why the index isn't on (did, queue, due, ord)? Then this new query should be possible without any slowdown.

Every extra indexed column adds a storage cost, and extra work when that row is updated. Anki 1 was a bit too liberal in its use of indexes, and I tried to reduce them in Anki 2 to the minimum possible.

Adding ord ASC here would allow us to fetch new cards in the default order faster, but we'd pay the extra column index cost for every card in the collection. We could perhaps work around it with a pair of conditional indexes (including ord if queue=0, and a separate index without ord where queue!=0), but I don't know what sort of performance implications that would entail. And unfortunately the index would only help in the default sort case - if the user wanted decreasing card positions for example, we either couldn't use the index, or we'd be giving the user decreasing ordinals as well.

Reply:
> In this case though we're relying on a separate index,

Without the `ord asc`? But I guess this is the same scenario as with the deck name sorting above. Maybe it's because it requires a tiny bit additional work to find the first entry fitting the `where` clause in the b-tree, whereas with arbitrary order SQLite just has to find any fitting entry and can expand from there.

Okay, pity about the extra sorting table, but that's probably not even noticeable in terms of performance.